.github/codeql/queries/unique-command-use.ql

```diff
@@ -0,0 +1,134 @@
+/**
+ * @name A VS Code command should not be used in multiple locations
+ * @kind problem
+ * @problem.severity warning
+ * @id vscode-codeql/unique-command-use
+ * @description Using each VS Code command from only one location makes
+ * our telemtry more useful because we can differentiate more user
+ * interactions and know which features of the UI users are using.
+ * To fix this alert, new commands will need to be made so that each one
+ * is only used from one location. The commands should share the same
+ * implementation so we do not introduce duplicate code.
+ * When fixing this alert, search the codebase for all other references
+ * to the commnad name. The location of the alert is an arbitrarily
+ * chosen usage of the command, and may not necessarily be the location
+ * that should be changed to fix the alert.
+ */
+
+ import javascript
+
+ /**
+  * The name of a VS Code command.
+  */
+ class CommandName extends string {
+   CommandName() { exists(CommandUsage e | e.getCommandName() = this) }
+ 
+   /**
+    * In how many ways is this command used. Will always be at least 1.
+    */
+   int getNumberOfUsages() { result = count(CommandUsage e | e.getCommandName() = this | e) }
+ 
+   /**
+    * Get the canonical first usage of this command, to use for the location
+    * of the alert. The implementation of this ordering of usages is arbitrary
+    * and the usage given may not be the one that should be changed when fixing
+    * the alert.
+    */
+   CommandUsage getFirstUsage() {
+     result.getCommandName() = this and
+     forall(CommandUsage e | e.getCommandName() = this |
+       e.getLocationOrdinal() >= result.getLocationOrdinal()
+     )
+   }
+ }
```

Comment:
QL suggestions that should preserve the meaning but hopefully make this more intuitive to maintain:
- Create a member predicate `CommandUsage getAUse() { result.getCommandName() = this}`
- `getNumberOfUsages` then becomes `count(this.getAUse())`
- `getFirstUsage` then becomes `result = max(CommandUsage use | use = this.getAUse() | use order by use.getLocationOrdinal())`
- Optional: drop `getLocationOrdinal` entirely and use `result = max(CommandUsage use | use = this.getAUse() | use order by use.getFile().getRelativePath(), use.getLocation().getStartLine(), use.getLocation().getStartColumn())`

Reply:
Thanks @adityasharad for the suggestions. That's a lot cleaner! ❤️